# Session 1 Data Structures

In [14]:
import pandas as pd
import os
folderName='data'
fileName='cities.xlsx'
fileExcel=os.path.join(folderName,fileName)
cv=pd.read_excel(fileExcel) 
cv

city   region  visits2013  visits2015
0        Bangkok     Asia       15.98       18.24
1          Dubai     Asia        9.89       14.26
2      Hong Kong     Asia        8.72        8.66
3       Istanbul   Europe       10.37       12.56
4   Kuala Lumpur     Asia        9.20       11.12
5         London   Europe       15.96       18.82
6       New York  America       11.52       12.27
7          Paris   Europe       13.92       16.06
8          Seoul     Asia         NaN       10.35
9      Singapore     Asia       11.75       11.88
10     Barcelona   Europe        8.41         NaN

 the most visited city in 2015:

In [2]:
MostVisited2015=cv[cv.visits2015==max(cv.visits2015)].city
MostVisited2015

5    London
Name: city, dtype: object

the least visited city in Asia in 2013:

In [3]:
LeastVisitedAisa2013=cv[cv.region=='Asia'].sort_values('visits2013',ascending=True).head(1).city
LeastVisitedAisa2013

2    Hong Kong
Name: city, dtype: object

the three least visited cities in Asia in 2015:

In [4]:
ThreeLeastVisitedAisa2015=cv[cv.region=='Asia'].sort_values('visits2015',ascending=True).head(3).city
ThreeLeastVisitedAisa2015

2       Hong Kong
8           Seoul
4    Kuala Lumpur
Name: city, dtype: object

the most visited city in Asia in 2013:

In [5]:
LeastVisitedAisa2015=cv[cv.region=='Asia'].sort_values('visits2015',ascending=True).head(1).city
LeastVisitedAisa2015

2    Hong Kong
Name: city, dtype: object

and the least visited city in Asia in 2013:

In [6]:
MostVisitedAsia2015=cv[cv.region=='Asia'].sort_values('visits2015',ascending=True).tail(1).city
MostVisitedAsia2015

0    Bangkok
Name: city, dtype: object

# Session 2: Control of Execution

using 'np.isnan':

In [10]:
from math import sqrt
import numpy as np

values=[9,None,'1000',-100, 144,-72]
for value in values: 
    if np.isnan(value): # condition1 
        print ('missing values as input')
        continue
    if isinstance(value, str): #condition2
        print ('string as input')
        continue
    if value < 0: # condition3
        print ('negative value as input')
        continue
    print (sqrt(value), 'is the root of ',value)  

3.0 is the root of  9


TypeError: ufunc 'isnan' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''

In [11]:
values=[9,np.nan,'1000',-100, 144,-72]
for value in values: 
    if np.isnan(value): # condition1 
        print ('missing values as input')
        continue
    if isinstance(value, str): #condition2
        print ('string as input')
        continue
    if value < 0: # condition3
        print ('negative value as input')
        continue
    print (sqrt(value), 'is the root of ',value)  

3.0 is the root of  9
missing values as input


TypeError: ufunc 'isnan' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''

In [12]:
values=[9,np.nan,3.6,-100, 144,-72]
for value in values: 
    if np.isnan(value): # condition1 
        print ('missing values as input')
        continue
    if isinstance(value, str): #condition2
        print ('string as input')
        continue
    if value < 0: # condition3
        print ('negative value as input')
        continue
    print (sqrt(value), 'is the root of ',value)  

3.0 is the root of  9
missing values as input
1.8973665961010275 is the root of  3.6
negative value as input
12.0 is the root of  144
negative value as input


There would be a TypeError that ufunc 'isnan' is not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''.

numpy.isnan test NaN, which is numerical data, but it can't be used to test other types such as strings.

Changing the order of conditions:

In [3]:
values=[9,None,np.nan, '1000',-100, 144,-72]
for value in values: 
    if value < 0: # condition3
        print ('negative value as input')
        continue
    if isinstance(value, str): #condition2
        print ('string as input')
        continue    
    if value==None: # condition1
        print ('missing values as input')
        continue
    print (sqrt(value), 'is the root of ',value)

3.0 is the root of  9


TypeError: '<' not supported between instances of 'NoneType' and 'int'

In [4]:
values=[9,None,np.nan, '1000',-100, 144,-72]
for value in values:    
    if value==None: # condition1
        print ('missing values as input')
        continue
    if value < 0: # condition3
        print ('negative value as input')
        continue
    if isinstance(value, str): #condition2
        print ('string as input')
        continue 
    print (sqrt(value), 'is the root of ',value) 

3.0 is the root of  9
missing values as input
nan is the root of  nan


TypeError: '<' not supported between instances of 'str' and 'int'

When condition 3 is not the last condition, there would be a TypeError that '<' is not supported between instances of 'NoneType' and 'int', or between 'str' and 'int'.

But it works when condition 3 is the last condition:

In [9]:
 values=[9,None,np.nan, '1000',-100, 144,-72]
for value in values: 
    if isinstance(value, str): #condition2
        print ('string as input')
        continue    
    if value==None: # condition1
        print ('missing values as input')
        continue
    if value < 0: # condition3
        print ('negative value as input')
        continue
    print (sqrt(value), 'is the root of ',value)  

3.0 is the root of  9
missing values as input
nan is the root of  nan
string as input
negative value as input
12.0 is the root of  144
negative value as input


Using isnull to test mising values:

In [10]:
import pandas as pd
values=[9,None,np.nan, '1000',-100, 144,-72]
for value in values: 
    if pd.isnull(value): # condition1
        print ('missing values as input')
        continue
    if isinstance(value, str): #condition2
        print ('string as input')
        continue
    if value < 0: # condition3
        print ('negative value as input')
        continue
    print (sqrt(value), 'is the root of ',value)    

3.0 is the root of  9
missing values as input
missing values as input
string as input
negative value as input
12.0 is the root of  144
negative value as input


using if and for

Import new data on people's age:

In [ ]:
import pandas as pd
import os

folderName='data'
fileName='people.xlsx'
fileExcel=os.path.join(folderName,fileName)
dataFromExcel=pd.read_excel(fileExcel) # table '1'
dataFromExcel

finding the youngest age, and inform who has that age.

In [11]:
numRows,numCols=dataFromExcel.shape

YoungestAge=1000
positionsOfYoungest=[]
countYoungest=0


for position in range(numRows):
    age = dataFromExcel.loc[position,'ages']
    if YoungestAge>age:
        YoungestAge=age


for position in range(numRows):
    age = dataFromExcel.loc[position,'ages']
    if YoungestAge==age:
        countYoungest=countYoungest+1
        positionsOfYoungest.append(position)

In [12]:
YoungestAge

25

In [13]:
dataFromExcel.iloc[positionsOfYoungest]

cod  ages state  education
2    A3    25    DC          3
7    A8    25    NY          2
19  A20    25    FL          1

Redo questions in session 1:

the most visited city in 2015:

In [15]:
numRows,numCols=cv.shape

mv2015=0
positionsOfmv2015=[]

for position in range(numRows):
    v2015 = cv.loc[position,'visits2015']
    if v2015>mv2015:
        mv2015=v2015

for position in range(numRows):
    v2015 = cv.loc[position,'visits2015']
    if v2015==mv2015:
        positionsOfmv2015.append(position)

cv.iloc[positionsOfmv2015].city

5    London
Name: city, dtype: object

the least visited in Aisa in 2013

In [16]:
numRows,numCols=cv.shape

lv2013a=100
positionsOflv2013a=[]

for position in range(numRows):
    region = cv.loc[position,'region']
    if region != 'Asia':
        continue
    v2013a = cv.loc[position,'visits2013']
    if v2013a<lv2013a:
        lv2013a=v2013a

for position in range(numRows):
    v2013a = cv.loc[position,'visits2013']
    if v2013a==lv2013a:
        positionsOflv2013a.append(position)

cv.iloc[positionsOflv2013a].city

2    Hong Kong
Name: city, dtype: object

the three least visited cities in Asia in 2015:

In [43]:
numRows,numCols=cv.shape

lv2015a1=100
positionsOflv2015a1=[]
lv2015a2=100
positionsOflv2015a2=[]
lv2015a3=100
positionsOflv2015a3=[]

for position in range(numRows):
    region = cv.loc[position,'region']
    if region != 'Asia':
        continue
    v2015a = cv.loc[position,'visits2015']
    if v2015a<lv2015a1:
        lv2015a1=v2015a

for position in range(numRows):
    v2015a = cv.loc[position,'visits2015']
    if v2015a==lv2015a1:
        positionsOflv2015a1.append(position)



for position in range(numRows):
    region = cv.loc[position,'region']
    if region != 'Asia':
        continue
    v2015a = cv.loc[position,'visits2015']
    if v2015a<lv2015a2 and v2015a>lv2015a1:
        lv2015a2=v2015a

for position in range(numRows):
    v2015a = cv.loc[position,'visits2015']
    if v2015a==lv2015a2:
        positionsOflv2015a2.append(position)



for position in range(numRows):
    region = cv.loc[position,'region']
    if region != 'Asia':
        continue
    v2015a = cv.loc[position,'visits2015']
    if v2015a<lv2015a3 and v2015a>lv2015a2 :
        lv2015a3=v2015a

for position in range(numRows):
    v2015a = cv.loc[position,'visits2015']
    if v2015a==lv2015a3:
        positionsOflv2015a3.append(position)

In [44]:
cv.iloc[positionsOflv2015a1].city 

2    Hong Kong
Name: city, dtype: object

In [45]:
cv.iloc[positionsOflv2015a2].city

8    Seoul
Name: city, dtype: object

In [46]:
cv.iloc[positionsOflv2015a3].city

4    Kuala Lumpur
Name: city, dtype: object

Using comphrehensions (with the values from rootsInDict):

In [ ]:
values=[9,25,49,-121]
rootsInDict={value:(sqrt(value) if value > 0 else None) for value in values}  

Creating a list of tuples:

In [47]:
ListofTuples=[value for value in (tuple((key,value)for key,value in rootsInDict.items()))]
ListofTuples

[(9, 3.0), (25, 5.0), (49, 7.0), (-121, None)]

Create a list with the first 10 even numbers:

In [48]:
list1=[0,2,4,6,8,10,12,14,16,18]

creating a list with each of those values to the power of two:

In [49]:
list2=[value**2 for value in list1]
list2

[0, 4, 16, 36, 64, 100, 144, 196, 256, 324]

creating another list with each of those same values with their square root:

In [50]:
list3=[sqrt(value)for value in list1]
list3

[0.0,
 1.4142135623730951,
 2.0,
 2.449489742783178,
 2.8284271247461903,
 3.1622776601683795,
 3.4641016151377544,
 3.7416573867739413,
 4.0,
 4.242640687119285]

creating a list of tuples combining the three lists, with each tuple holding the triplet: (number, numer to power of two, squared root of number.

In [51]:
list4=list(zip(list1,list2,list3))
list4

[(0, 0, 0.0),
 (2, 4, 1.4142135623730951),
 (4, 16, 2.0),
 (6, 36, 2.449489742783178),
 (8, 64, 2.8284271247461903),
 (10, 100, 3.1622776601683795),
 (12, 144, 3.4641016151377544),
 (14, 196, 3.7416573867739413),
 (16, 256, 4.0),
 (18, 324, 4.242640687119285)]

Creating a dictionary of dictionaries with the three lists created initially in the previous exercise. The dict of dicts should be named 'numPowRoot', and if you write: numPowRoot[10], your output should be {'power2': 100, 'root2': 3.1622776601683795}

In [52]:
numPowRoot ={value: {'power2': value ** 2, 'root2' : sqrt(value)} for value in list1}
numPowRoot

{0: {'power2': 0, 'root2': 0.0},
 2: {'power2': 4, 'root2': 1.4142135623730951},
 4: {'power2': 16, 'root2': 2.0},
 6: {'power2': 36, 'root2': 2.449489742783178},
 8: {'power2': 64, 'root2': 2.8284271247461903},
 10: {'power2': 100, 'root2': 3.1622776601683795},
 12: {'power2': 144, 'root2': 3.4641016151377544},
 14: {'power2': 196, 'root2': 3.7416573867739413},
 16: {'power2': 256, 'root2': 4.0},
 18: {'power2': 324, 'root2': 4.242640687119285}}

# Session 3: Building Functions

Creating the function riseToPowerPlus, which gives a good answer even when the power is negative.

In [55]:
def riseToPowerPlus(base,exponent=2): 
    result=1
    if exponent > 0:
        for time in range(1,exponent+1): 
            result=result*base
    elif exponent < 0:
        for time in range(1,-exponent):
            result = 1/(result*base)
    else: 
        result = 1
    return(result)

In [56]:
riseToPowerPlus(base=9,exponent=-2)

0.1111111111111111

Change the function ’factors’to reduce the amount of iterations in the for loop:

In [15]:
def factors2(number):
    factorsList=[] 
    for i in range(1,int((sqrt(number)+1))):
        if number % i == 0:
            factorsList.append(int(i))
            factorsList.append(int(number/i))
    return factorsList 

In [16]:
factors2(20)

[1, 20, 2, 10, 4, 5]

a function that does not accept negative values or zeros as input and returns a message if an invalid value was input.

In [17]:
def factors3(number):
    factorsList=[] 
    if number <= 0:
        print('sorry, nevative value or zero input are not accepted.')
    if number >0:
        for i in range(1,int((sqrt(number)+1))):
            if number % i == 0:
                factorsList.append(int(i))
                factorsList.append(int(number/i))
        return factorsList 

Or

In [19]:
def factors3(number):
    factorsList=[] 
    if number == 0:
        print('zero as input, empty list.')
    try:
        for i in range(1,int((sqrt(number)+1))):
            if number % i == 0:
                factorsList.append(int(i))
                factorsList.append(int(number/i))
        return factorsList 
    except ValueError:
        print('negative value input is not accepted.')

In [20]:
factors3(0)

zero as input,empty list


[]

a function that accepts only positive integer values; and returns a message if an invalid valid was input.

In [24]:
def factors(number):
    factorsList=[] 
    if number >0 and type(number)== int:
        for i in range(1,int((sqrt(number)+1))):
            if number % i == 0:
                factorsList.append(int(i))
                factorsList.append(int(number/i))
        return factorsList 
    else:
        print('sorry, only positive integer inputs are accepted.')

or

In [22]:
def factors(number):
    factorsList=[] 
    if isinstance(number,int) and number > 0: 
        for i in range(1,int((sqrt(number)+1))):
            if number % i == 0:
                factorsList.append(int(i))
                factorsList.append(int(number/i))
        return factorsList 
    else:
        print('sorry, input invalid')

In [27]:
factors(8.0)

sorry, only positive integer inputs are accepted.


a function 'resistant' to missing values of any kind:

In [ ]:
import pandas as pd
def factors(number):
    factorsList=[] 
    if pd.notnull(number):
        if isinstance(number,int) and number > 0: 
            for i in range(1,int((sqrt(number)+1))):
                if number % i == 0:
                    factorsList.append(int(i))
                    factorsList.append(int(number/i))
            return factorsList 
        else:
            print('invalid value as input')
    else:
        print('missing value as input')

In [ ]:
factors(0)

a function that reads two lists and returns a data frame with those lists and extra columns with their sum, difference, multiplication and division.

In [ ]:
def extra_columns(list_a, list_b):
    import pandas as pd
    twonumbers = zip(list_a,list_b)
    sumlist =[]
    diflist =[]
    mullist =[]
    divlist =[]
    for i in twonumbers:
        sumval = i[0] + i[1]
        sumlist.append(sumval)
        difval = i[0] - i[1]
        diflist.append(difval)
        mulval = i[0] * i[1]
        mullist.append(mulval)
        divval = i[0]/i[1]
        divlist.append(divval)
    extra_dic ={'List  A': list_a, 'List B': list_b, 'SUM': sumlist, 'DIF': diflist, 'MUL': mullist, 'DIV': divlist }
    return pd.DataFrame(extra_dic)

or

In [ ]:
def extra(list_a, list_b):
    df = pd.DataFrame({'listA':list_a,'listB':list_b}, columns =['listA','listB'])
    sum_list = df.apply(lambda x: x[0]+x[1], axis = 1)
    dif_list = df.apply(lambda x: x[0] - x[1], axis = 1)
    mul_list =df.apply(lambda x: x[0]*x[1], axis = 1)
    div_list =df.apply(lambda x: x[0]/x[1], axis = 1)
    extra_dic ={'List  A': list_a, 'List B': list_b, 'SUM': sum_list, 'DIF': dif_list, 'MUL': mul_list, 'DIV': div_list }
    return pd.DataFrame(extra_dic)


In [ ]:
extra_columns(factors(10),factors(4))

to get double() to work as it did with R vectors

In [28]:
def double(x):
    doublelist=[]
    for i in x:
        doublelist.append(i*2)
    return doublelist

In [29]:
myList=[1,2,3]
double(myList)

[2, 4, 6]

understand the following function:

In [30]:
def strangerF(x):
    answer=0
    change=True
    for value in x:
        if (change):
            answer+=value
        else:
            answer-=value
    
        change= not change 

    return answer

strangerF sums the first value, minus second value, sums the third, minus the fourth... for example, if I have a list=[1,2,3,4,5], strangerF gives the answer of "+1-2+3-4+5"

In [36]:
testList=[1,2,3,4,5]
strangerF(testList)

3

In [31]:
folderName='data'
fileName='dataforstrangeF.csv'
fileF=os.path.join(folderName,fileName)
dataforstrangeF=pd.read_csv(fileF) 
dataforstrangeF

vals1  vals2  vals3
0     A      9    100
1     B      1      1
2     C      1      1
3     A      1      1
4     B      1      1
5     C      1      1
6     A      1      1
7     B      1      1
8     C      1      1

strangerF can be applied to the vals2 and vals3 columns because values in these two columns are nemeric.
But apply, mapply and filter don't work.